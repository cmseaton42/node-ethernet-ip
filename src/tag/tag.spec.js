const Tag = require("./index");
const { Types } = require("../enip/cip/data-types");

describe("Tag Class", () => {
    describe("New Instance", () => {
        it("Throws Error on Invalid Inputs", () => {
            const fn = (tagname, prog, type = Types.UDINT) => {
                return () => new Tag(tagname, prog, type);
            };

            expect(fn(1234)).toThrow();
            expect(fn("hello")).not.toThrow();
            expect(fn("someTag", "prog", 0x31)).toThrow();
            expect(fn("someTag", "prog", Types.EPATH)).not.toThrow();
            expect(fn("someTag", "prog", 0xc1)).not.toThrow();
        });
    });

    describe("Tagname Validator", () => {
        it("Accepts and Rejects Appropriate Inputs", () => {
            const fn = test => Tag.isValidTagname(test);

            expect(fn("_sometagname")).toBeTruthy();
            expect(fn(12345)).toBeFalsy();
            expect(fn(null)).toBeFalsy();
            expect(fn(undefined)).toBeFalsy();
            expect(fn(`hello${311}`)).toBeTruthy();
            expect(fn("hello.how3")).toBeTruthy();
            expect(fn("randy.julian.bubbles")).toBeTruthy();
            expect(fn("a.b.c")).toBeTruthy();
            expect(fn("1.1.1")).toBeFalsy();
            expect(fn({ prop: "value" })).toBeFalsy();
            expect(fn("fffffffffffffffffffffffffffffffffffffffff")).toBeFalsy();
            expect(fn("ffffffffffffffffffffffffffffffffffffffff")).toBeTruthy();
            expect(fn("4hello")).toBeFalsy();
            expect(fn("someTagArray[12]")).toBeTruthy();
            expect(fn("someTagArray[1a]")).toBeFalsy();
            expect(fn("hello[f]")).toBeFalsy();
            expect(fn("someOtherTag[0]a")).toBeFalsy();
            expect(fn("tagname")).toBeTruthy();
            expect(fn("tag_with_underscores45")).toBeTruthy();
            expect(fn("someTagArray[0]")).toBeTruthy();
        });
    });

    describe("Read Message Generator Method", () => {
        it("Generates Appropriate Buffer", () => {
            const tag1 = new Tag("tag", null, Types.DINT);
            const tag2 = new Tag("tag", null, Types.BOOL);
            const tag3 = new Tag("tag", null, Types.REAL);
            const tag4 = new Tag("tag", null, Types.SINT);
            const tag5 = new Tag("tag", null, Types.INT);

            expect(tag1.generateReadMessageRequest()).toMatchSnapshot();
            expect(tag2.generateReadMessageRequest()).toMatchSnapshot();
            expect(tag3.generateReadMessageRequest()).toMatchSnapshot();
            expect(tag4.generateReadMessageRequest()).toMatchSnapshot();
            expect(tag5.generateReadMessageRequest()).toMatchSnapshot();
        });
    });

    describe("Write Message Generator Method", () => {
        it("Generates Appropriate Buffer", () => {
            const tag1 = new Tag("tag", null, Types.DINT);
            const tag2 = new Tag("tag", null, Types.BOOL);
            const tag3 = new Tag("tag", null, Types.REAL);
            const tag4 = new Tag("tag", null, Types.SINT);
            const tag5 = new Tag("tag", null, Types.INT);

            expect(tag1.generateWriteMessageRequest(100)).toMatchSnapshot();
            expect(tag2.generateWriteMessageRequest(true)).toMatchSnapshot();
            expect(tag3.generateWriteMessageRequest(32.1234)).toMatchSnapshot();
            expect(tag4.generateWriteMessageRequest(4)).toMatchSnapshot();
            expect(tag5.generateWriteMessageRequest(-10)).toMatchSnapshot();
        });
    });



    describe("keepAlive parameter", () => {
        it("should allow a number input", () => {
            const testTag = new Tag("testkeepalive", undefined, undefined, 10);
            expect(testTag).toBeInstanceOf(Tag);
        });

        it("should throw an error on non-number types", () => {
            expect(() => {
                new Tag("testkeepalive", undefined, undefined, "apple");
            }).toThrowError("Tag expected keepAlive of type <number> instead got type <string>");
        });

        it("should throw an error if keepAlive is less than 0", () => {
            expect(() => {
                new Tag("testkeepalive", undefined, undefined, -20);
            }).toThrowError("Tag expected keepAlive to be greater than 0, got -20");
        });
    });
});
