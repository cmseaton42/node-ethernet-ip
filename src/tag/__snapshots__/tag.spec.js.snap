// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag Class Read Message Generator Method Generates Appropriate Buffer 1`] = `
Object {
  "data": Array [
    76,
    3,
    145,
    3,
    116,
    97,
    103,
    0,
    1,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`Tag Class Read Message Generator Method Generates Appropriate Buffer 2`] = `
Object {
  "data": Array [
    76,
    3,
    145,
    3,
    116,
    97,
    103,
    0,
    1,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`Tag Class Read Message Generator Method Generates Appropriate Buffer 3`] = `
Object {
  "data": Array [
    76,
    3,
    145,
    3,
    116,
    97,
    103,
    0,
    1,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`Tag Class Read Message Generator Method Generates Appropriate Buffer 4`] = `
Object {
  "data": Array [
    76,
    3,
    145,
    3,
    116,
    97,
    103,
    0,
    1,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`Tag Class Read Message Generator Method Generates Appropriate Buffer 5`] = `
Object {
  "data": Array [
    76,
    3,
    145,
    3,
    116,
    97,
    103,
    0,
    1,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`Tag Class Read Message Generator Method Generates Appropriate Buffer 6`] = `
Object {
  "data": Array [
    76,
    3,
    145,
    3,
    116,
    97,
    103,
    0,
    1,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`Tag Class Read Message Generator Method Generates Appropriate Buffer 7`] = `
Object {
  "data": Array [
    76,
    4,
    145,
    3,
    116,
    97,
    103,
    0,
    40,
    0,
    1,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`Tag Class Read Message Generator Method Generates Appropriate Buffer 8`] = `
Object {
  "data": Array [
    76,
    5,
    145,
    3,
    116,
    97,
    103,
    0,
    40,
    0,
    40,
    0,
    1,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`Tag Class Read Message Generator Method Generates Appropriate Buffer 9`] = `
Object {
  "data": Array [
    76,
    6,
    145,
    3,
    116,
    97,
    103,
    0,
    40,
    0,
    40,
    0,
    40,
    0,
    1,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`Tag Class Write Message Generator Method Generates Appropriate Buffer 1`] = `
Object {
  "data": Array [
    77,
    3,
    145,
    3,
    116,
    97,
    103,
    0,
    196,
    0,
    1,
    0,
    100,
    0,
    0,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`Tag Class Write Message Generator Method Generates Appropriate Buffer 2`] = `
Object {
  "data": Array [
    77,
    3,
    145,
    3,
    116,
    97,
    103,
    0,
    193,
    0,
    1,
    0,
    1,
  ],
  "type": "Buffer",
}
`;

exports[`Tag Class Write Message Generator Method Generates Appropriate Buffer 3`] = `
Object {
  "data": Array [
    77,
    3,
    145,
    3,
    116,
    97,
    103,
    0,
    202,
    0,
    1,
    0,
    93,
    126,
    0,
    66,
  ],
  "type": "Buffer",
}
`;

exports[`Tag Class Write Message Generator Method Generates Appropriate Buffer 4`] = `
Object {
  "data": Array [
    77,
    3,
    145,
    3,
    116,
    97,
    103,
    0,
    194,
    0,
    1,
    0,
    4,
  ],
  "type": "Buffer",
}
`;

exports[`Tag Class Write Message Generator Method Generates Appropriate Buffer 5`] = `
Object {
  "data": Array [
    77,
    3,
    145,
    3,
    116,
    97,
    103,
    0,
    195,
    0,
    1,
    0,
    246,
    255,
  ],
  "type": "Buffer",
}
`;

exports[`Tag Class Write Message Generator Method Generates Appropriate Buffer 6`] = `
Object {
  "data": Array [
    78,
    3,
    145,
    3,
    116,
    97,
    103,
    0,
    4,
    0,
    1,
    0,
    0,
    0,
    255,
    255,
    255,
    255,
  ],
  "type": "Buffer",
}
`;

exports[`Tag Class Write Message Generator Method Generates Appropriate Buffer 7`] = `
Object {
  "data": Array [
    77,
    4,
    145,
    3,
    116,
    97,
    103,
    0,
    40,
    0,
    196,
    0,
    1,
    0,
    99,
    0,
    0,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`Tag Class Write Message Generator Method Generates Appropriate Buffer 8`] = `
Object {
  "data": Array [
    77,
    5,
    145,
    3,
    116,
    97,
    103,
    0,
    40,
    0,
    40,
    0,
    196,
    0,
    1,
    0,
    99,
    0,
    0,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`Tag Class Write Message Generator Method Generates Appropriate Buffer 9`] = `
Object {
  "data": Array [
    77,
    6,
    145,
    3,
    116,
    97,
    103,
    0,
    40,
    0,
    40,
    0,
    40,
    0,
    196,
    0,
    1,
    0,
    99,
    0,
    0,
    0,
  ],
  "type": "Buffer",
}
`;
